doc/developer/design/20230717_alter_set_cluster.md

```diff
@@ -0,0 +1,88 @@
+# Moving source and sinks
+
+Issue: [#17417](https://github.com/MaterializeInc/materialize/issues/17417)
+
+## Context
+
+Materialize's API today allows users to specify a target cluster for sources
+and sinks on creation, but does not allow for changing the cluster at a later
+time. To change the cluster, users have to drop and recreate the object.
+
+This is problematic:
+* Sources almost always have downstream dependencies that need to be recreated,
+  too.
+* Sinks require users to recreate the topic in the downstream Kafka broker.
+
+## Goals
+
+* Users chan change the cluster on which a source or sink runs.
+
+## Non-Goals
+
+* Mixing compute and storage workload on the same cluster.
+
+## Overview
+
+We'll add support for the following declarative source and sink management
+commands to Materialize:
+
+```sql
+-- Move a source to cluster `clstr`:
+ALTER SOURCE src SET CLUSTER clstr;
+
+-- Move a sink to cluster `c2`:
+ALTER SINK snk SET CLUSTER c2;
+```
+
+Note that this is _intentionally_ lacking parens
+(`ALTER SOURCE src SET (CLUSTER = clstr)`;).
+This is for consistency with the analogous PostgreSQL command
+`ALTER TABLE ... SET {SCHEMA|TABLESPACE} {schema|tblspc}`.
+The rule is that only parameters configured in the `WITH` block go inside the
+parens; parameters that have dedicated syntax, like schemas, tablespaces, and
+clusters, get dedicated `ALTER ... SET` syntax too.
+
+## Detailed description
+
+The `ALTER SOURCE` and `ALTER SINK` statements will learn a new option `SET CLUSTER clstr`:
+* The cluster name `clstr` must match an existing cluster.
+* The usual constraints apply. At the moment, the cluster must either be empty
+  or only host sources and sinks. The replication factor must be 0 or 1.
+* Clusters can be managed or unmanaged.
+
+When the command updates the cluster, Materialize will immediately drop the
+existing source or sink and recreate it in the target cluster.
+
+> **Warning**
+> 
+> In the initial implementation, changing the cluster associated will result in
+> downtime while the source or sink recreate.
+> 
+> We hope to change the behavior in a future release so that no downtime is
+> incurred, but this will require smarter source and sink scheduling.
+
+Users may not schedule compute objects together with sources and sinks.
+Attempting to schedule incompatible objects will result in an error like the
+following:
+
+```sql
+ALTER SOURCE src SET CLUSTER compute_cluster;
+ERROR: cannot alter source cluster to cluster containing indexes or materialized views
+```
+
+Users may not schedule sources or sinks on clusters with an incompatible
+replication factor. Attempting to do so will result in an error like the
+following:
+
+```sql
+ALTER SOURCE src SET CLUSTER cluster;
+ERROR: cannot source source cluster to cluster with more than one replica 
+```
+
```

Comment:
Can you add a discussion for how this will work with linked clusters? Will it be permissible to migrate a source with a `SIZE` to a source with a `CLUSTER`? What about the reverse?

Reply:
Added a comment. Going from a linked cluster to an explicit cluster is simple. I'm not sure we want to allow the reverse!
* How far are we on deprecating linked clusters? This would be an opportunity to speed this up a but.
* Switching back to a lined cluster would mean that we either remember a previously specified `SIZE` value, or need to specify on the `ALTER` statement. In the first case, we could offer `RESET` to do the same thing, but this seems highly specific.

Reply:
> How far are we on deprecating linked clusters? This would be an opportunity to speed this up a but.

Depends on the cluster unification work! As soon as sources can default to being created in the current cluster (https://github.com/MaterializeInc/materialize/issues/17420) we can think about deprecating linked clusters (https://github.com/MaterializeInc/materialize/issues/17422).

> Switching back to a lined cluster would mean that we either remember a previously specified SIZE value, or need to specify on the ALTER statement. In the first case, we could offer RESET to do the same thing, but this seems highly specific.

I was envisioning `ALTER SOURCE ... SET (SIZE = ...)` as the way you can migrate a source back to a linked cluster. (Rather than trying to remember its original size.) But I'd be 👍🏽 on skipping this unless proven absolutely necessary by a user, in the spirit of moving towards deprecating linked clusters.

Reply:
> But I'd be 👍🏽 on skipping this unless proven absolutely necessary by a user, in the spirit of moving towards deprecating linked clusters.

+1, I can't think of a scenario where users would take a source/sink back to a linked cluster. My impression is that these are used mostly because all the examples in the documentation use `SIZE`, and we've refrained from replacing them so far (see #19888). The most common scenario is for users to create sources/sinks using this option while exploring on their own, but then drop and recreate them in a dedicated ingestion cluster once they get in touch with FE and start to grok best practices.

Since we're now creating an easy path to migration from linked clusters, I'd wait to change the example until the cluster unification work is complete. WDYT, @sjwiesman?